File: solver/behavior.js
// The few seconds after a clearance is issued, which are not idle time.
//
// Shared because replay.js needs to do exactly what the solver does before it
// can claim to have measured the same thing. Its in-session probe — solve a
// challenge in a fresh context, navigate again in that context — is the
// measurement the whole verdict turns on, and it was re-navigating the instant
// the interstitial cleared, which is precisely the case the dwell below exists
// to avoid. A "the zone challenges everything" verdict taken that way could just
// be a cookie read too early.

// Real-user behavior between challenge solve and cookie capture. CF samples
// mouse/scroll/dwell events for the first few seconds after issuance and
// uses them to score the cookie. A cookie captured cold dies under load
// within seconds; one captured after 3-5 s of believable activity holds up.
export async function simulateHumanBehavior(page) {
  // Initial settle pause (RUM beacons fire here).
  await sleep(rand(900, 1600));

  // Mouse move into a believable region.
  try {
    await page.mouse.move(rand(220, 1100), rand(180, 600), { steps: 12 });
  } catch {}
  await sleep(rand(280, 600));

  // First scroll - slow, smooth.
  try {
    await page.evaluate(
      (y) => window.scrollBy({ top: y, behavior: "smooth" }),
      rand(280, 600)
    );
  } catch {}
  await sleep(rand(800, 1300));

  // Second mouse move + smaller scroll.
  try {
    await page.mouse.move(rand(400, 1500), rand(300, 800), { steps: 10 });
  } catch {}
  try {
    await page.evaluate(
      (y) => window.scrollBy({ top: y, behavior: "smooth" }),
      rand(150, 380)
    );
  } catch {}
  await sleep(rand(700, 1100));

  // Dispatch a focus/blur event, simulating tab switching back.
  try {
    await page.evaluate(() => {
      window.dispatchEvent(new Event("focus"));
      document.body && document.body.click();
    });
  } catch {}
  await sleep(rand(500, 900));

  // Final dwell - lets __cf_bm settle and any post-load JS finish.
  await sleep(rand(600, 1100));
}

File: internal/solver/behavior.go
package solver

import (
	"context"
	"math/rand"
	"strconv"
	"time"

	"github.com/JSInvasor/Gohttp-clientfingerprintemulateandfastest/internal/cdp"
)

// The few seconds after a clearance is issued, which are not idle time.
//
// Cloudflare samples mouse, scroll and dwell events for the first few seconds
// after issuance and scores the cookie on them. A cookie captured cold dies
// under load within seconds; one captured after 3-5s of believable activity
// holds up. So this runs before the jar is read, not after.
//
// Every step is best-effort. A page that navigated out from under a mouse move
// is the challenge clearing, which is the outcome being waited for — failing the
// solve over it would discard the cookie the behaviour was meant to strengthen.
func simulateHumanBehavior(ctx context.Context, tab *cdp.Tab) {
	// Initial settle pause: RUM beacons fire here.
	if !nap(ctx, 900, 1600) {
		return
	}

	_ = tab.MouseMove(ctx, randRange(220, 1100), randRange(180, 600), 12)
	if !nap(ctx, 280, 600) {
		return
	}

	// Smooth rather than instant: a scroll that jumps produces one event where a
	// wheel produces a series.
	_ = tab.Evaluate(ctx, scrollBy(randRange(280, 600)), nil)
	if !nap(ctx, 800, 1300) {
		return
	}

	_ = tab.MouseMove(ctx, randRange(400, 1500), randRange(300, 800), 10)
	_ = tab.Evaluate(ctx, scrollBy(randRange(150, 380)), nil)
	if !nap(ctx, 700, 1100) {
		return
	}

	// A focus event and a click, the way a tab switched back to behaves.
	_ = tab.Evaluate(ctx, `(() => {
		window.dispatchEvent(new Event("focus"));
		if (document.body) document.body.click();
	})()`, nil)
	if !nap(ctx, 500, 900) {
		return
	}

	// Final dwell: lets __cf_bm settle and any post-load JS finish.
	nap(ctx, 600, 1100)
}

func scrollBy(y float64) string {
	return `window.scrollBy({top: ` + strconv.Itoa(int(y)) + `, behavior: "smooth"})`
}

// nap sleeps for a jittered interval, reporting false if the context ended
// first — the caller's budget outranks the choreography.
func nap(ctx context.Context, minMS, maxMS int) bool {
	d := time.Duration(minMS+rand.Intn(maxMS-minMS+1)) * time.Millisecond
	timer := time.NewTimer(d)
	defer timer.Stop()
	select {
	case <-ctx.Done():
		return false
	case <-timer.C:
		return true
	}
}

func randRange(lo, hi float64) float64 {
	return lo + rand.Float64()*(hi-lo)
}
